framework.cpp

```diff
@@ -0,0 +1,403 @@
+#ifndef _METRIC_FRAMEWORK_CPP
+#define _METRIC_FRAMEWORK_CPP
+
+#include <cmath>
+#include <vector>
+//#include <map>
+
+#include "3rdparty/blaze/Blaze.h"
+
+
+namespace metric {
+
+
+
+
+// --------
+// common type trait metafunctions
+
+
+
+template <typename>
+struct determine_container_type  // checks whether container is STL container (1) or Blaze vector (2)
+{
+    constexpr static int code = 0;
+};
+
+template <template <typename, typename> class Container, typename ValueType, typename Allocator>
+struct determine_container_type<Container<ValueType, Allocator>>
+{
+    constexpr static int code = 1; // STL
+};
+
+template <template <typename, bool> class Container, typename ValueType, bool F>
+struct determine_container_type<Container<ValueType, F>>
+{
+    constexpr static int code = 2; // blaze::DynamicVector
+};
+
+template <template <typename, int, int, int, int, int> class Container, typename ValueType, int _Rows, int _Cols, int _Options, int _MaxRows, int _MaxCols>
+struct determine_container_type<Container<ValueType, _Rows, _Cols, _Options, _MaxRows, _MaxCols>>
+{
+    constexpr static int code = 3; // Eigen::Array
+};
+
+
+template<typename C, int = determine_container_type<C>::code>
+struct determine_val_t  // determines type of element both for STL containers and Blaze vectors
+{
+    using type = void;
+};
+
+template<typename C>
+struct determine_val_t<C, 1>
+{
+    using type = typename C::value_type;
+};
+
+template<typename C>
+struct determine_val_t<C, 2>
+{
+    using type = typename C::ElementType;
+};
+
+template<typename C>
+struct determine_val_t<C, 3>
+{
+    using type = typename C::Scalar;
+};
+
+
+
+// --------
+// base classes
+
+
+template <typename rec_t>
+class MetricBase { // all metrics to be derived from
+
+public:
+
+    using Rec_T = rec_t;
+    using Val_T = determine_val_t<Rec_T>;
+
+    MetricBase<rec_t>() {}
+
+    double operator () (const rec_t & a, const rec_t & b) { // to be overridden in derived metrics
+        return 0;
+    }
+
+private:
+
+};
+
+
+
+
+template <typename Metric = MetricBase<double>>
+class Algorithm {
+
+public:
+
+    //using Rec_T = rec_t;
+    //using Val_T = determine_val_t<Rec_T>;
+
+    Algorithm<Metric>(Metric metric_) : metric(metric_) {}
```

Comment:
some algorithms asks for more than one metric, this will make it asymmetric, what do you think?

Reply:
perhaps we could have multiple base algorithms?

Reply:
several metrics used together in single algorithm? If so, we can use variadic templates as a common API, but, again, it's more complex.. What are these algorithms? maybe we can tackle them somehow specially

Yes, we can use nultiple base algorithms. Moreover, we can do not use inheritance at all, simply hold the standard. In the current code there is no need in inheritance, i added it just as an example that shows its possibility

Reply:
@maxfilippov for example MGC